### ISSUE TITLE
AssemblyLoadContext issue, needs better documentation or default behavior for AssemblyLoadContext

### ISSUE THREAD

**Issue by Seabizkit** [dotnet/runtime#73900]
Running in to all sorts of issues no matter what i try.. i have tied many many different things.
I updated all LIB to .net 6, to rule out any funnies, getting the same issue.

What is the correct way to use AssemblyLoadContext?
I want main app and the plugin to share lib "Paradox.Component.Facade" and nothing else.
I have followed the MS examples and many many others, this code tester below is based off one of MS example.

EXCEPTION:

System.Reflection.ReflectionTypeLoadException: 'Unable to load one or more of the requested types.
Method 'ExcuteAsync' in type 'Paradox.Component.FtpDownloadPlugin.Plugin' from assembly 'Paradox.Component.FtpDownloadPlugin, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.'

Project structure

  --PLUGIN INTERFACE
    Paradox.Component.Facade
         - IParadoxComponent
       
  --SHARE  LIB WHERE AssemblyLoadContext is;
    Paradox.ComponentCode
    REF->Paradox.Component.Facade
        - PluginLoadContext
  
  --TESTER
    Paradox.Plugin.Tester
    REF->Paradox.Component.Facade
    REF->Paradox.ComponentCode
          - Program (code below)
  
  --PLUGIN EXAMPLE
    Paradox.Component.FtpDownloadPlugin
    REF->Paradox.Component.Facade

project file for Paradox.Component.FtpDownloadPlugin

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
	  <TargetFramework>.net6</TargetFramework>
	  <ImplicitUsings>enable</ImplicitUsings>
	  <Nullable>enable</Nullable>
	  <EnableDynamicLoading>true</EnableDynamicLoading>
  </PropertyGroup>
	
	<ItemGroup>
		<PackageReference Include="Microsoft.Extensions.Hosting" Version="6.0.0" />
		<PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="6.0.0" />
	</ItemGroup>
	
  <ItemGroup>
<ProjectReference Include="..\..\Paradox.Component.Facade.Std2\Paradox.Component.Facade.Std2.csproj">
	<Private>false</Private>
</ProjectReference>
	
<ProjectReference Include="..\..\..\Paradox.OnDiskHelperWrapper\Paradox.OnDiskHelperWrapper.csproj" />
<ProjectReference Include="..\..\..\Paradox.Common.Std2\Paradox.Common.Std2.csproj" />
<ProjectReference Include="..\..\..\Paradox.OnDiskHelper.Facade\Paradox.OnDiskHelperWrapper.Facade.csproj" />
<ProjectReference Include="..\..\..\Paradox.FileHelper.Facade\Paradox.FileHelperWrapper.Facade.csproj" />
<ProjectReference Include="..\..\..\Paradox.FileHelper\Paradox.FileHelperWrapper.csproj" />
<ProjectReference Include="..\..\..\Data\Paradox.Data.Importer.Facade\Paradox.Data.Importer.Facade.csproj" />
<ProjectReference Include="..\..\..\Data\Paradox.Data.Importer\Paradox.Data.Importer.net5.csproj" />
<ProjectReference Include="..\..\..\Ftp\Paradox.FtpFactory.Std2\Paradox.AppFtpFactory.Std2.csproj" />
<ProjectReference Include="..\..\..\Ftp\Paradox.FtpCommon.Std2\Paradox.FtpCommon.Std2.csproj" />
  </ItemGroup>
</Project>
```

code for plugin
```
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Paradox.AppFtpFactory.Std2;
using Paradox.Component.Facade.Std2;
using Paradox.Component.Facade.Std2.Models;
using Paradox.Data.Importer.Facade.Interfaces.Repo.BuilderRepo;
using Paradox.Data.Importer.Facade.Interfaces.Repo.ImpoterRepo;
using Paradox.Data.Importer.net5.Extensions;
using Paradox.FtpCommon.Std2;
using Microsoft.Extensions.Configuration;
using Paradox.FileHelperWrapper.Facade;
using Paradox.FileHelperWrapper;
using Paradox.OnDiskHelperWrapper.Facade;
using Paradox.Common.Std2.ResponseCode;
using Paradox.OnDiskHelperWrapper.Extensions;

namespace Paradox.Component.FtpDownloadPlugin;

  public class Plugin : IParadoxComponent
  {
      public  Guid IdentifferName => new Guid("C9BEB8BE-D031-4A30-B88B-7E9D3894D65E");
      public string Name => "Paradox FTP Download";
      public string ShortName => "Ftp Download";
      public string Author => "Paradox";
      public string Description => "Utility for downloading files using FTP via already define Importer ftp connection.";
      public string ImageIcon => "fa-cloud-download";
      public string ImageUrl => "";
  
      private IServiceProvider SetupService(ExecutingContext e, IConfiguration configuration)
      {
          var host = DIBuilder.Start(e, services => {
  
              var loggerFactory = LoggerFactory.Create(builder => { builder.AddConsole(); });
              ILogger logger = loggerFactory.CreateLogger<Plugin>();
  
              //database
              services.AddImporterDatabase(configuration, logger);
              //repo core
              services.AddImpoterCoreRepos(logger);
              services.AddImpoterBuilderRepos(logger);
  
              services.AddStorageLocation(configuration, logger);
  
              services.AddSingleton<IFtpFactory, FtpFactory>();
              services.AddSingleton<IFileHelper, FileHelper>();
  
          });
          return host;
      }
  
      public async Task<ComponentResponse> ExcuteAsync(ExecutingContext ec, CancellationToken ct = default)
      {
          var response = new ComponentResponse();
  
          try
          {
              var configuration = DIBuilder.SetConfigation(ec);
              var host = SetupService(ec, configuration);
  
              using (var scope = host.CreateScope())
              {
                  var services = scope.ServiceProvider;
  
                  var ftpConnectionRepo = services.GetRequiredService<IFtpConnectionRepo>();
                  var ftpFactory = services.GetRequiredService<IFtpFactory>();
  
                  var builderTaskRunRepo = services.GetRequiredService<IBuilderTaskRunRepo>();
                  var onDiskHelper = services.GetRequiredService<IOnDiskHelper>();
  
                  var fileHelper = services.GetRequiredService<IFileHelper>();
  
                  var instance = new ComponentProvider(
                                      ftpConnectionRepo,
                                          ftpFactory,
                                          builderTaskRunRepo,
                                          onDiskHelper,
                                          fileHelper,
                                          ec.Logger
                                  );
  
                  response = await instance.ExecuteAsync(ec, ct);
              }
          }
          catch (OperationCanceledException ex)
          {
              // Task was cancelled
              await ec.Logger.LogErrorAsync($"Cancelled {ex.Message}", ex);
              throw;
          }
          catch (Exception ex)
          {
              await ec.Logger.LogErrorAsync($"Failed {ex.Message}", ex);
              throw;
          }
  
          return response;
      }
  }

```

```
public class PluginLoadContext : AssemblyLoadContext
    {

        private AssemblyDependencyResolver _resolver;

        public PluginLoadContext() : base(isCollectible: true)
        { }

        public PluginLoadContext(string pluginPath, params Type[] sharedTypes) : base(isCollectible: true)
        {
            _resolver = new AssemblyDependencyResolver(pluginPath);
        }

        protected override Assembly Load(AssemblyName assemblyName)
        {
            if (_resolver != null)
            {
                string assemblyPath = _resolver.ResolveAssemblyToPath(assemblyName);
                if (assemblyPath != null)
                {
                    return LoadFromAssemblyPath(assemblyPath);
                }
            }
            return null;
        }

        protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
        {
            string libraryPath = _resolver.ResolveUnmanagedDllToPath(unmanagedDllName);
            if (libraryPath != null)
            {
                return LoadUnmanagedDllFromPath(libraryPath);
            }

            return IntPtr.Zero;
        }

        public IEnumerable<T> GetImplementations<T>(string assebblylocation)
        {
            var type = typeof(T);
            //var typeAssName = type.Assembly.GetName();
            var assName = System.Reflection.AssemblyName.GetAssemblyName(assebblylocation);
            var assembly = this.LoadFromAssemblyName(assName);
            
            //EXCEPTION IS THROWN HERE when trying to get types.
            var types = assembly.GetTypes().ToList();
            //var filteredAssemblys = loadedAssemblies.Where(x => x.GetName().Name.StartsWith("Paradox")).ToList();

            List<Type> other = new List<Type>();
            try
            {
                //  var types = filteredAssemblys.SelectMany(a => a.GetTypes()).ToList();
                //The code that causes the error goes here.
                if (type.IsAbstract)
                {
                    other = types.Where(myType => myType.IsClass && !myType.IsAbstract
                                            && myType.IsSubclassOf(type)
                                    ).ToList();
                }
                else
                {
                    other = types.Where(t => type.IsAssignableFrom(t)).ToList();
                }




            }
            catch (ReflectionTypeLoadException ex)
            {
                StringBuilder sb = new StringBuilder();
                foreach (Exception exSub in ex.LoaderExceptions)
                {
                    sb.AppendLine(exSub.Message);
                    FileNotFoundException exFileNotFound = exSub as FileNotFoundException;
                    if (exFileNotFound != null)
                    {
                        if (!string.IsNullOrEmpty(exFileNotFound.FusionLog))
                        {
                            sb.AppendLine("Fusion Log:");
                            sb.AppendLine(exFileNotFound.FusionLog);
                        }
                    }
                    sb.AppendLine();
                }
                string errorMessage = sb.ToString();
                //Display or log the error based on your application.
            }


            return other

                .Select(t => Activator.CreateInstance(t))
                .Cast<T>();

        }
    }
```

code .net6 console

```
// See https://aka.ms/new-console-template for more information
using Paradox.Component.Facade.Std2;
using Paradox.ComponentCode.Code;
using System.Reflection;

Console.WriteLine("Plugin  tester");

//look in the pluginFolder..
string[] pluginPaths = new string[]
{
    @"Paradox.Component.FtpDownloadPlugin\Paradox.Component.FtpDownloadPlugin.dll",
};

List<IParadoxComponent> commands = new List<IParadoxComponent>();

foreach (var item in pluginPaths)
{
    Assembly pluginAssembly = LoadPlugin(item);
    var comane = CreateCommands(pluginAssembly);
    commands.Add(comane);
}

static Assembly LoadPlugin(string relativePath)
{
    //var path = Path.Combine(directoryName);
    var path = @"C:\Projects\Paradox.Importer\src\_plugins";
    string root = Path.GetFullPath(path);

    string pluginLocation = Path.GetFullPath(Path.Combine(root, relativePath.Replace('\\', Path.DirectorySeparatorChar)));
    Console.WriteLine($"Loading commands from: {pluginLocation}");
    PluginLoadContext loadContext = new PluginLoadContext(pluginLocation);

    var instanve = loadContext.GetImplementations<IParadoxComponent>(pluginLocation);

    var name = AssemblyName.GetAssemblyName(pluginLocation);
    return loadContext.LoadFromAssemblyName(name);
}

static IParadoxComponent CreateCommands(Assembly assembly)
{
    var types = assembly.GetTypes();
    foreach (Type type in types)
    {
        if (typeof(IParadoxComponent).IsAssignableFrom(type))
        {
            IParadoxComponent? result = Activator.CreateInstance(type) as IParadoxComponent;
            if (result != null)
            {
                return result;
            }
        }
    }
    string availableTypes = string.Join(",", types.Select(t => t.FullName));
    throw new ApplicationException(
        $"Can't find any type which implements ICommand in {assembly} from {assembly.Location}.\n" +
        $"Available types: {availableTypes}");
    
}
```

**Comment by vitek-karas** [comment]
The most likely cause for the exception is that the class `ExecutingContext` comes from an assembly which is loaded both by the host as well as the plugin (so two copies). The assembly which defines `IParadoxComponent` probably has additional dependencies. All of those must be loaded only by the host, not by the plugins.

This typically means that the plugin project should reference the assembly which defines the interface like this:
```C#
<ItemGroup>
    <ProjectReference Include="..\PluginBase\PluginBase.csproj">
        <Private>false</Private>
        <ExcludeAssets>runtime</ExcludeAssets>
    </ProjectReference>
</ItemGroup>
```
(as seen here: https://docs.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support#simple-plugin-with-no-dependencies)

Note that it's important to have both `Private=false` (which works on ProjectReferences) as well as `ExcludeAssets=runtime` (which works on PackageReferences).

To validate this, make sure that all the assemblies which are referenced by the `Paradox.Component.Facade` are NOT included in the output of the plugin build. This can also mean that if the plugin references one of these directly (PackageReference) it needs to use the same mechanism as above to not include the assembly in its output.

Reason for the exception:
The host loads the `IParadoxComponent` type which references the `ExecutingContext` and thus that gets loaded into the host. Now the plugin references `IParadoxComponent` which is resolved to the same assembly as the one used by the host, but the `ExecutingContext` is resolved into a new copy of its assembly into the load context of the plugin. And so the plugin's implementation of `ExecuteAsync` doesn't match the signature of the interface method as seen by the host, because the parameter type's don't match (same name, but different assembly).

**Comment by vitek-karas** [comment]
> I want main app and the plugin to share lib "Paradox.Component.Facade" and nothing else.

I think the correct way to look at this is that the main app and plugin must share the `Paradox.Component.Facade` AND everything that assembly depends on. (or more precise they have to share all types which occur in the plugin interface, no matter where they come from).

**Comment by Seabizkit** [comment]
@vitek-karas I believe that helped a lot!, thanks

you were correct I had another lib shared in the assembly which defines IParadoxComponent.
I have since removed that and things are better but I'm still not quite where I would like to be. 
Aka new issue grrrr

Thanks you! if you dnt mind, could you help me understand my next issue.

I removed all other libs from: Paradox.Component.Facade (one which holds the IParadoxComponent) it now loads and makes sense.

To give context on the next thing its giving me a hard time about;

When my main app loads, it may not have a ddl loaded as there is nothing which calls it directly.
 aka its called via reflection, not sure what the best way to handle this is but i have.. is code which calls EnsureAssemblyLoaded


Code 1:
```
  //todo: dll optermisations needed here for loading.....cache the loaded dlls
    var listAss = new List<string> { actoinReq.ImplementationDll, actoinReq.InterfaceDll };
    var assemblies = new List<Assembly>(listAss.Count());
    foreach (var item in listAss) 
    {
        _logger.LogInformation($"Getting assembly Type {item}");
        **var assembly = ReflectionHelpers.GetAssembly(item);**
        assemblies.Add(assembly);
    }
```
inside ReflectionHelpers.GetAssembly(item);

Code 2:
```
public static class ReflectionHelpers
    {
        public static Assembly GetAssembly(string namepace)
        {
            //https://dotnetcoretutorials.com/2020/07/03/getting-assemblies-is-harder-than-you-think-in-c/
            EnsureAssemblyLoaded(namepace);

            **//this is returning two dll which match the same name...**
            var assemblies = **AppDomain.CurrentDomain.GetAssemblies()**
                                    .Where(x => x.GetName().Name == namepace)
                                    .OrderBy(x=>x.FullName)
                                    .ToList();

            var assembly = assemblies.SingleOrDefault();

            return assembly;
        }
}
```
Code 3:
```
   public static void EnsureAssemblyLoaded(string assemblyName)
        {
            var buildDirectory = GetApplicationPath();
            var phyicalPath = Path.Combine(buildDirectory,$"{assemblyName}.dll");
            var assembly = AssemblyName.GetAssemblyName(phyicalPath);
            **AppDomain.CurrentDomain.Load(assembly);**
        }
```

on first run of another process its fine, but afterwards aka after its loaded

Code: 4
```
var pluginInfo = await _taskComponentHelper.EnsureLocalPluginAsync(item.BuilderComponentId.Value);
var pluginLocation = Path.GetFullPath(Path.Combine(pluginInfo.Path, pluginInfo.Name));
var loadContext = new PluginLoadContext(pluginLocation);
var assName = System.Reflection.AssemblyName.GetAssemblyName(pluginLocation);
var assembly = loadContext.LoadFromAssemblyName(assName);
var types = assembly.GetTypes();
var pluginUI = ReflectionHelpers.GetClassInstanceWhichDerivesFromBaseFromTypes<BaseComponentUI>(types).Single();
var plugin = ReflectionHelpers.GetClassInstanceWhichImplemtsInterfaceFromTypes<IParadoxComponent>(types).Single();
```
there is another process which will call the other case again...Code 1:
but now it will have two dll of the same name (Paradox.ComponentCode.dll).

This  (Paradox.ComponentCode.dll) is common between the main app and the plugins.
But the way I would like it to work is that the plugin keeps all its dll separate even if the name are the same.

Why is it that it now has two dlls and how would I prevent it from doing this.
Aka anything which it loads from Plugin folder should not be part of AppDomain.CurrentDomain

I hope that makes sense.

**Comment by vitek-karas** [comment]
Please don't use AppDomains or even think about them :wink: in .NET (Core). Unlike .NET Framework, there's only ever one AppDomain - and so it effectively acts "global". So for example `AppDomain.CurrentDomain.GetAssemblies` will return all assemblies loaded the app, regardless of load context. So if the same assembly is loaded into the default load context and then also by a plugin into a plugin load context, it will be listed twice in that collection.

If you need to enumerate assemblies in the default load context, you can use `AssemblyLoadContext.Default.Assemblies`.

> Aka anything which it loads from Plugin folder should not be part of AppDomain.CurrentDomain

This is basically "thinking about it wrong" - there's only one AppDomain - ever, so everything belongs to it.
*Ideally we would remove AppDomains completely, but that would break too much code*

If you're into the "Reading" phase, definitely look at this: https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/overview
It should have hopefully reasonable description of the system, how it works, and how to do things.

**Comment by Seabizkit** [comment]
Thanks! 
I'm still not sure what I would replace that section with, let me explain...
.net 6, I need to get a class via reflection, as the system hasn't called it yet, it hasn't been loaded yet.
I need to load it, the assembly where this class is so i can get an instance via reflection.
The class in question here can be loaded* as if it was part of the app aka not a plugin isolation.

I presume AppDomain.CurrentDomain, is where all other dlls would normally get added if there were reference.
Like this site suggests....
//https://dotnetcoretutorials.com/2020/07/03/getting-assemblies-is-harder-than-you-think-in-c/
```
AppDomain.CurrentDomain.GetAssemblies(); // Does not return SomeAssembly as it hasn't been called yet. 
SomeAssembly.SomeClass.SomeMethod();
AppDomain.CurrentDomain.GetAssemblies(); // Will now return SomeAssembly. 
```

Could you give me a run down of what is expected, let me try and explain with flow.

1. Start app, runs things, not important
2. An endpoint is called
3. That endpoint creates and instance of MyFoo* via reflection, this class (MyFoo) can come from main app,
 but its not reference yet like the above snip, so need add it.(assembly name: Paradox.ComponentCode) 
(how to get it so I can have an instance).
4.Currently I am doing "AppDomain.CurrentDomain.Load(assembly);" which like you say will add it to well everything aka the App.
5. This gets me access to it to ask for it via reflection.
6. MyFoo starts a process which will create plugin's using AssemblyLoadContext using specific folder for each one.
7. it seems like that when AssemblyLoadContext is used and asked to load  which also uses (Paradox.ComponentCode)  it adds another (Paradox.ComponentCode) to the global (AppDomain). (u are correct my understanding of how this works was not correct)
I havnt checked yet but i assume it sees these as different due to the load location.
8. In my code step 2, which is called again it will fail as it see the same name twice.

Sorry I know I repeating myself and the issue, you say that I should avoid using AppDomain in a sense.

So when step 2 is called I should create an AssemblyLoadContext aswell, seems odd when its folder location is the main app.
trying to think how else this can be done, but from you ur sayiny, there would be no way... as when i query for it, it would then give me two.

am I mean to create AssemblyLoadContext without a path in this instance.? when is the usage of that applicable.
 lol im so confused about this last part.

Side question:
not tested yet but i presume that when you call unload on AssemblyLoadContext it would then remove it from AppDomain  then.

in a nut shell;
Need to load type from unloaded dll (Paradox.ComponentCode)
This DLL is also used in plugin but is (private to that one) well isolated well that what I want it to be.
I want to be able to query for the type from Paradox.ComponentCode.

After the plugin runs there will be two and that would mean i would need to query by path or something. what would be the best approach here, I dnt quite get how AssemblyLoadContext adds to AppDomain and not quite sure how to query for type if AppDomain is going to have two.

I think the answer is use AssemblyLoadContext, do i just set the resolving path to the main app....
the thing is currently i do not store the path, would I have to set this or will it know? aka set the AssemblyLoadContext to the app bin folder, or are there some clever defaults I'm not understanding here.

currently the value I have saved is : Paradox.ComponentCode (no path)
basically in my head I'm not seeing how I get way from two copies without setting a path on the AssemblyLoadContext.
I feel like im missing something here.

**Comment by vitek-karas** [comment]
Please read the docs here, they should explain the concepts much better than I can do here: https://docs.microsoft.com/en-us/dotnet/core/dependency-loading/overview

Very short version (for .NET 6):
* Only one AppDomain - so all assemblies are in it, no matter where they're loaded
* Each assembly is loaded into one AssemblyLoadContext (ALC)
* Each ALC can load only one assembly of any given simple name
* Two separate ALCs can load the same assembly - in this case the two instances are separate and runtime sees them as different assemblies (even if loaded from the same exact file on disk). In this case AppDomain will contain both assemblies (as it contains all of them)
* The main application assembly is loaded into the Default ALC (`AssemblyLoadContext.Default`)
* Due to resolution mechanisms (see the link above), this means that typically all of the assemblies which the app directly depends on are loaded into Default ALC as well
* The .NET Framework assemblies (`System.*`) should only be loaded into Default ALC, loading them into other ALCs as well will typically cause weird behavior

> I need to get a class via reflection, as the system hasn't called it yet, it hasn't been loaded yet.
> I need to load it, the assembly where this class is so i can get an instance via reflection.

If the assembly in question is part of the application (the app was compiled with a reference to it), then you don't need to explicitly load it, the resolution will figure it out. All you need to do is to call `Type.GetType("MyType, MyAssemblyName")` and under the hood this will call `Assembly.Load("MyAssemblyName")` and find the type in that assembly for you.

If the assembly is not part of the app, then you have to tell the system where to find it (note that it's not enough for the file to be in the same directory, .NET 6 will not automatically look around on disk for assemblies). For this `Assembly.Load` won't work as it doesn't say where to find it. For this, use `AssemblyLoadContext.LoadFromAssemblyPath` - if you want to load the assembly "as if part of the app", then you would probably call `AssemblyLoadContext.Default.LoadFromAssemblyPath` as that will load it into the default ALC (where the rest of the app is).

**Comment by Seabizkit** [comment]
You have helped so much! thanks you for your time.
